bluemira/structural/transformation.py

```diff
@@ -0,0 +1,232 @@
+    d = np.hypot(a, b)
+    # TODO: Why does the less intuitive, simpler algebra form work better??
+    # https://ocw.mit.edu/courses/civil-and-environmental-engineering/1-571-structural-analysis-and-control-spring-2004/readings/connor_ch5.pdf  # noqa
+    if np.isclose(a, 0) and np.isclose(b, 0):
+        dcm = np.array(
+            [[0.0, 0.0, -np.sign(c)], [0.0, 1.0, 0.0], [np.sign(c), 0.0, 0.0]]
+        )
+    else:
+        dcm = np.array([[a, -b / d, -a * c / d], [b, a / d, -b * c / d], [c, 0, d]]).T
+
+    return dcm
+
+
+def _direction_cosine_matrix_debugging(dx, dy, dz, debug=False):
+    """
+    Slow, ugly, safe
+    """
+    dcm = 0
+    u = np.array([dx, dy, dz])
+    x_local = u / np.linalg.norm(u)
+
+    x_global = np.array([1.0, 0, 0])
+    y_global = np.array([0, 1.0, 0])
+    z_global = np.array([0, 0, 1.0])
+    globa = np.array([x_global, y_global, z_global])
+
+    if x_local[0] == 1:
+        # corresponds to the global coordinate system
+        dcm = np.eye(3)
+        if debug:
+            return dcm, globa
+        else:
+            return dcm
+    if x_local[0] == -1:
+        # corresponds to the mirrored coordinate system
+        dcm = np.eye(3)
+        dcm[0, 0] = -1
+        local = globa.copy()
+        local[0][0] = -1
+        if debug:
+            return dcm, local
+        else:
+            return dcm
+    if abs(x_local[1]) == 1:
+        # corresponds to a local y-vector sitting on the global x-vector
+        # (rotation about z-axis)
+        cos_theta = 0
+        sin_theta = np.sign(x_local[1])
+        dcm = np.array(
+            [[cos_theta, -sin_theta, 0], [sin_theta, cos_theta, 0], [0, 0, 1]]
+        )
+        if debug:
+            local = np.array([[0, sin_theta, 0], [-sin_theta, 0, 0], [0, 0, 1]])
+            return dcm, local
+        else:
+            return dcm
+    if x_local[2] == 1:
+        # corresponds to a local z-vector sitting on the global x-vector
+        # rotation about the y-axis
+        cos_theta = 0
+        sin_theta = -1
+        dcm = np.array(
+            [[cos_theta, 0, sin_theta], [0, 1, 0], [-sin_theta, 0, cos_theta]]
+        )
+        if debug:
+            local = np.array([[0, 0, 1], [0, 1, 0], [-1, 0, 0]])
+            return dcm, local
+        else:
+            return dcm
+
+    y_local = np.array([0, 1.0, 0])
+    y_local -= np.dot(x_local, y_local) * x_local
+    y_local /= np.linalg.norm(y_local)
+    z_local = np.cross(x_local, y_local)
+
+    local = np.array([x_local, y_local, z_local])
+
+    if not isinstance(dcm, int):
+        if debug:
+            return dcm, local
+        else:
+            return dcm
+    c11 = np.dot(x_global, x_local)
+    c12 = np.dot(x_global, y_local)
+    c13 = np.dot(x_global, z_local)
+    c21 = np.dot(y_global, x_local)
+    c22 = np.dot(y_global, y_local)
+    c23 = np.dot(y_global, z_local)
+    c31 = np.dot(z_global, x_local)
+    c32 = np.dot(z_global, y_local)
+    c33 = np.dot(z_global, z_local)
```

Comment:
einsum may be quicker here if you get round to it

Reply:
not used atm. never quite got round to working out why I couldnt get this one to work